### PR TITLE
add super documents

### PR DESCRIPTION
**Pull request by eva24752460**
add:

utils_super_documents (Method for the tfidf super documents)
model_tfidf_cos (Model for predictions TF-IDF Cosine Similarity)
model_tfidf_super_documents_naive (Model for predictions TF-IDF naive with super documents)

modify (add the option to use the 'super documents' method) :

model_tfidf_dense
model_tfidf_gbt
model_tfidf_lgbm
model_tfidf_sgdc
model_tfidf_svm

### CODE REVIEW

File: gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_cos.py [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+            x_test (?): Array-like or sparse matrix, shape = [n_samples]
+        Kwargs:
+            return_proba (bool): If the function should return the probabilities instead of the classes (Keras compatibility)
+        Returns:
+            (np.ndarray): Array, shape = [n_samples]
+        '''
+        if return_proba:
+            return self.predict_proba(x_test)
+        else:
+            return self.compute_scores(x_test)
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_cos.py:211]: Not intuitive, 'scores' and return predictions ?

File: gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_cos.py [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+            x_test (?): Array-like or sparse matrix, shape = [n_samples]
+        Returns:
+            (np.ndarray): Array, shape = [n_samples]
+        '''
+        x_test = np.array([x_test]) if isinstance(x_test, str) else x_test
+        x_test = np.array(x_test) if isinstance(x_test, list) else x_test
+
+        chunk_size = 5000
+        vec = self.pipeline.transform(x_test).astype(np.float16)
+        self.matrix_train = self.matrix_train.astype(np.float16)
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_cos.py:248]: It should be done in fit, not in predict

File: gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_cos.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+        vec_size = math.ceil((vec.shape[0])/chunk_size)
+        train_size = math.ceil((self.matrix_train.shape[0])/chunk_size)
+        array_predicts = np.array([], dtype='int')
+        for vec_row in range(vec_size):
+            block_vec = vec[vec_row*chunk_size:(vec_row+1)*chunk_size]
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_cos.py:253]: pep8

File: gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_cos.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+            list_cosine = []
+            for train_row in range(train_size):
+                block_train = self.matrix_train[train_row*chunk_size:(train_row+1)*chunk_size]
+                cosine = cosine_similarity(block_train, block_vec).astype(np.float16)
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_cos.py:257]: can probably be improved !

File: gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_super_documents_naive.py [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+    '''Model for predictions via TF-IDF + Naive'''
+
+    _default_name = 'model_tfidf_super_documents_naive'
+
+    def __init__(self, tfidf_count_params: Union[dict, None] = None, tfidf_transformer_params: Union[dict, None] = None, 
+                 multiclass_strategy: Union[str, None] = None, **kwargs):
+        '''Initialization of the class (see ModelPipeline & ModelClass for more arguments)
+
+        Kwargs:
+            tfidf_transformer_params (dict): Parameters for the tfidf TfidfTransformer
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_super_documents_naive.py:54]: fix kwargs order

File: gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_super_documents_naive.py [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+        if tfidf_transformer_params is None:
+            tfidf_transformer_params = {}
+        self.tfidf = TfidfTransformerSuperDocuments(**tfidf_transformer_params)
+
+        if tfidf_count_params is None:
+            tfidf_count_params = {}
+        self.tfidf_count = CountVectorizer(**tfidf_count_params)
+
+        self.multiclass_strategy = multiclass_strategy
+        self.matrix_train = csr_matrix((0,0))
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_super_documents_naive.py:83]: Can probably be set to None

File: gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_super_documents_naive.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+        self.array_target = np.array([])
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_super_documents_naive.py:84]: Can probably be set to None

File: gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_super_documents_naive.py [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+
+        # Can't do multi-labels / multi-classes
+        if not self.multi_label:
+            # If not multi-classes : no impact
+            if multiclass_strategy == 'ovr':
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_super_documents_naive.py:89]: Change this to : ... in ['ovo', 'ovr']: ...

File: gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_super_documents_naive.py [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+                raise ValueError("The TFIDF Naive can't do ovr")
+            elif multiclass_strategy == 'ovo':
+                raise ValueError("The TFIDF Naive can't do ovo")
+            else:
+                self.pipeline = Pipeline([('tfidf_count', self.tfidf_count),('tfidf', self.tfidf)])
+
+    def save(self, json_data: Union[dict, None] = None) -> None:
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_super_documents_naive.py:96]: Must be at the end of this file, check other classes

File: gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_super_documents_naive.py [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+        Kwargs:
+            json_data (dict): Additional configurations to be saved
+        '''
+        # Save model
+        if json_data is None:
+            json_data = {}
+
+        # No need to save the parameters of the pipeline steps, it is already done in ModelPipeline
+        json_data['multiclass_strategy'] = self.multiclass_strategy
+        json_data['with_super_documents'] = self.with_super_documents
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_super_documents_naive.py:108]: Not useful if already done in model_pipeline

File: gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_cos.py [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+        Kwargs:
+            json_data (dict): Additional configurations to be saved
+        '''
+        # Save model
+        if json_data is None:
+            json_data = {}
+
+        # No need to save the parameters of the pipeline steps, it is already done in ModelPipeline
+        json_data['multiclass_strategy'] = self.multiclass_strategy
+        json_data['with_super_documents'] = self.with_super_documents
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_cos.py:102]: Not useful if already done in model_pipeline

File: gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_super_documents_naive.py [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+        json_data['classes_'] = self.tfidf.classes_ if hasattr(self.tfidf, 'classes_') else None
+
+        np.savetxt(os.path.join(self.model_dir, 'matrix_train.csv'), self.matrix_train.toarray(), delimiter=";")
+        np.savetxt(os.path.join(self.model_dir, 'array_target.csv'), self.array_target, delimiter=";", fmt="%s")
+
+        # Save
+        super().save(json_data=json_data)
+
+    def reload_from_standalone(self, **kwargs):
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_super_documents_naive.py:117]: Must be at the end of this file, check other classes

File: gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_super_documents_naive.py [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+from sklearn.feature_extraction.text import CountVectorizer
+
+from {{package_name}} import utils
+from {{package_name}}.models_training import utils_models
+from {{package_name}}.models_training.model_pipeline import ModelPipeline
+from {{package_name}}.models_training.utils_super_documents import TfidfTransformerSuperDocuments
+
+
+class ModelTfidfSuperDocumentsNaive(ModelPipeline):
+    '''Model for predictions via TF-IDF + Naive'''
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_super_documents_naive.py:45]: Add more explanations ?

File: gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_super_documents_naive.py [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+            x_test (?): Array-like or sparse matrix, shape = [n_samples]
+        Kwargs:
+            return_proba (bool): If the function should return the probabilities instead of the classes (Keras compatibility)
+        Returns:
+            (np.ndarray): Array, shape = [n_samples]
+        '''
+        if return_proba:
+            return self.predict_proba(x_test)
+        else:
+            return self.compute_scores(x_test)
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_super_documents_naive.py:219]: Not intuitive, 'scores' and return predictions ?

File: gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_super_documents_naive.py [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+            # Format ['a', 'b', 'c', 'a', ..., 'b']
+            # Transform to "proba"
+            transform_dict = {col: [0. if _ != i else 1. for _ in range(len(self.list_classes))] for i, col in enumerate(self.list_classes)}
+            probas = np.array([transform_dict[x] for x in preds])
+        else:
+            raise ValueError("The TFIDF Naive does not support multi label")
+        return probas
+
+    @utils.trained_needed
+    def compute_scores(self, x_test) -> np.ndarray:
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_super_documents_naive.py:243]: Not intuitive, 'scores' and return predictions ?

File: gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_super_documents_naive.py [hunk 16]
```diff
@@ -0,0 +1,261 @@
...
+        if return_proba:
+            return self.predict_proba(x_test)
+        else:
+            return self.compute_scores(x_test)
+
+    @utils.data_agnostic_str_to_list
+    @utils.trained_needed
+    def predict_proba(self, x_test, **kwargs) -> np.ndarray:
+        '''Predicts the probabilities on the test set
+        - /!\\ THE MODEL NAIVE DOES NOT RETURN PROBABILITIES, HERE WE SIMULATE PROBABILITIES EQUAL TO 0 OR 1 /!\\ -
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/model_tfidf_super_documents_naive.py:225]: Can't this return a score ?

File: gabarit/template_nlp/nlp_project/package_name/models_training/utils_super_documents.py [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+import numpy as np
+import pandas as pd
+
+from sklearn.feature_extraction.text import TfidfVectorizer, TfidfTransformer
+
+
+class TfidfTransformerSuperDocuments(TfidfTransformer):
+    '''TfidfTransformer for super documents'''
+
+    def get_super_documents_count_vectorizer(self, x_train, y_train):
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/utils_super_documents.py:32]: type hint ?

File: gabarit/template_nlp/nlp_project/package_name/models_training/utils_super_documents.py [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+# - TfidfVectorizerSuperDocuments -> TfidfVectorizer for super documents
+
+import numpy as np
+import pandas as pd
+
+from sklearn.feature_extraction.text import TfidfVectorizer, TfidfTransformer
+
+
+class TfidfTransformerSuperDocuments(TfidfTransformer):
+    '''TfidfTransformer for super documents'''
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/utils_super_documents.py:30]: Add more explanations on super documents ?

File: gabarit/template_nlp/nlp_project/package_name/models_training/utils_super_documents.py [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+
+        Args:
+            x_train (csr_matrix): shape = [n_samples, n_term]
+            y_train (?): Array-like, shape = [n_samples, n_targets]
+        Returns:
+            result(np.array): array, shape = [n_targets, n_term]
+            y_train(np.array): array, shape = [n_targets]
+        '''
+        index_array = np.array([np.where(y_train==x)[0] for x in np.unique(y_train)], dtype=object)
+        result = np.array([[sum(y) for y in x_train[x,:].transpose().toarray()] for x in index_array])
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/utils_super_documents.py:43]: pep8

File: gabarit/template_nlp/nlp_project/package_name/models_training/utils_super_documents.py [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+        '''
+        raw_super_documents, target_super_documents = self.get_super_documents_count_vectorizer(raw_documents, y)
+        self.fit(raw_super_documents, y)
+        return self.transform(raw_documents)
+
+
+class TfidfVectorizerSuperDocuments(TfidfVectorizer):
+    '''TfidfVectorize for super documents'''
+
+    def get_super_documents(self, x_train, y_train):
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/utils_super_documents.py:63]: Type Hint

File: gabarit/template_nlp/nlp_project/package_name/models_training/utils_super_documents.py [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+            y_train (?): Array-like, shape = [n_samples, n_targets]
+        Returns:
+            super_train(np.array): array, shape = [n_targets]
+            y_train(np.array): array, shape = [n_targets]
+        '''
+        x_train = pd.Series(x_train, name='x_train')
+        y_train = pd.Series(y_train, name='y_train')
+
+        df_train = pd.concat([x_train, y_train], axis=1)
+        super_train = df_train.groupby(y_train.name).agg({x_train.name:lambda sentence: ' '.join((sentence))})
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/utils_super_documents.py:77]: pep8

File: gabarit/template_nlp/nlp_project/package_name/models_training/utils_super_documents.py [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+        return np.array(super_train[x_train.name]), np.array(super_train.index)
+
+    def fit(self, raw_documents, y=None):
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/utils_super_documents.py:80]: type hint

File: gabarit/template_nlp/nlp_project/package_name/models_training/utils_super_documents.py [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+            y_train (?): Array-like, shape = [n_samples, n_targets]
+        Returns:
+            result(np.array): array, shape = [n_targets, n_term]
+            y_train(np.array): array, shape = [n_targets]
+        '''
+        index_array = np.array([np.where(y_train==x)[0] for x in np.unique(y_train)], dtype=object)
+        result = np.array([[sum(y) for y in x_train[x,:].transpose().toarray()] for x in index_array])
+        return result, np.unique(y_train)
+
+    def fit_transform(self, raw_documents, y=None):
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/utils_super_documents.py:46]: type hint

File: gabarit/template_nlp/nlp_project/package_name/models_training/utils_super_documents.py [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+        '''Trains and transform the model with super documents
+
+        Args:
+            raw_documents (?): Array-like, shape = [n_samples, n_targets]
+            y (?): Array-like, shape = [n_samples, n_targets]
+        Returns:
+            (csr_matrix): matrix, shape = [n_samples, n_term]
+        '''
+        self.fit(raw_documents, y)
+        return self.transform(raw_documents)
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/utils_super_documents.py:103]: Add if __name__ == '__main__': .... error

File: gabarit/template_nlp/nlp_project/package_name/models_training/utils_super_documents.py [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+            raw_documents (?): Array-like, shape = [n_samples, n_targets]
+            y (?): Array-like, shape = [n_samples, n_targets]
+        Returns:
+            TfidfVectorizerSuperDocuments
+        '''
+        raw_super_documents, target_super_documents = self.get_super_documents(raw_documents, y)
+        X = super().fit(raw_super_documents)
+        return self
+
+    def fit_transform(self, raw_documents, y=None):
```
Comment (LexABzH) [inline gabarit/template_nlp/nlp_project/package_name/models_training/utils_super_documents.py:93]: type hint

**Review by LexABzH — CHANGES_REQUESTED** [review]
- Lots of comments in files changes
- Functional tests must be added
- Train script must import new models
- 0_reload ... mus tbe updated to manage new models

File: gabarit/template_nlp/nlp_project/package_name/models_training/model_pipeline.py [hunk 26]
```diff
@@ -40,11 +40,12 @@ class ModelPipeline(ModelClass):
     # Not implemented :
     # -> reload
 
-    def __init__(self, pipeline: Union[Pipeline, None] = None, **kwargs) -> None:
+    def __init__(self, pipeline: Union[Pipeline, None] = None, with_super_documents: bool=False, **kwargs) -> None:
```
[the remark this answers is not in the document]
Comment (gsolard) [reply gabarit/template_nlp/nlp_project/package_name/models_training/model_pipeline.py:43]: I agree. But then, one has to individually change each of the tfidf model. One solution would be to add a new class ModelTfidf wich is a child of ModelPipeline and a parent of all model_tfidf_... (except  model_tfidf_dense of course). So we do as you want:
1) change each individual model (but it is less practical to maintain)
2) add another class ModelTfidf (the best solution in my opinion)

[the remark this answers is not in the document]
Comment (LexABzH) [reply gabarit/template_nlp/nlp_project/package_name/models_training/model_pipeline.py:43]: We can leave it like this for the moment. We'll see later if we have pipeline models without tfidf.

**Review by LexABzH — APPROVED** [review]

**Pull request closed without merging**